src/promptflow/promptflow/_sdk/_record_storage.py

```diff
@@ -0,0 +1,228 @@
+        elif isinstance(record_file, Path):
+            self._record_file = record_file.resolve()
+        else:
+            return
+
+        if self._record_file.parts[-1] != RecordStorage._standard_record_name:
+            record_folder = self._record_file / RecordStorage._standard_record_folder
+            self._record_file = record_folder / RecordStorage._standard_record_name
+        else:
+            record_folder = self._record_file.parent
+
+        self._record_file_str = str(self._record_file)
+
+        # cache folder we could create if not exist.
+        if not record_folder.exists():
+            record_folder.mkdir(parents=True, exist_ok=True)
+
+        # if file exist, load file
+        if self._record_file.exists():
+            self._load_file()
+
+    def _write_file(self) -> None:
+        file_content = self.cached_items.get(self._record_file_str, None)
+
+        if file_content is not None:
+            with open(self.record_file, "w+", encoding="utf-8") as fp:
+                json.dump(self.cached_items[self._record_file_str], fp, indent=4)
+
+    def _load_file(self) -> None:
+        local_content = self.cached_items.get(self._record_file_str, None)
+        if not local_content:
+            if not self.record_file.exists():
+                return
+            with open(self.record_file, "r", encoding="utf-8") as fp:
+                self.cached_items[self._record_file_str] = json.load(fp)
+
+    def get_record(self, input_dict: OrderedDict) -> object:
+        """
+        get_record Get record from local storage.
+
+        :param input_dict: input dict of critical AOAI inputs
+        :type input_dict: OrderedDict
+        :raises RecordFileMissingException: Record file not exist
+        :raises RecordItemMissingException: Record item not exist in record file
+        :return: original output of node run
+        :rtype: object
+        """
+        hash_value: str = hashlib.sha1(str(input_dict).encode("utf-8")).hexdigest()
+        current_saved_records: Dict[str, str] = self.cached_items.get(self._record_file_str, None)
+        if current_saved_records is None:
+            raise RecordFileMissingException(
+                f"This should except earlier, record file not found in folder {self.input_path}."
+            )
+        saved_output = current_saved_records.get(hash_value, None)
+        if saved_output is None:
+            raise RecordItemMissingException(
+                f"Record item not found in folder {self.input_path}.\n" f"values: {json.dumps(input_dict)}\n"
+            )
+
+        # not all items are reserved in the output dict.
+        output = saved_output["output"]
+        output_type = saved_output["output_type"]
+        if isinstance(output, str) and output_type != "str":
+            output_convert = json.loads(output)
+            return output_convert
+        else:
+            return output
+
+    def set_record(self, input_dict: OrderedDict, output: object) -> None:
```

Comment:
1. I'm not pretty sure if a `OrderedDict` is suitable for annotation, maybe a simple `Dict` is OK.
2. If one annotation is `object`, maybe we can simply ignore this?

Reply:
OK